### DISCUSSION TITLE
The strategies for `Reliable Long-form Transcription` in whisper.cpp differs from OpenAI's Whisper

### DISCUSSION THREAD

**Discussion by bobqianic** [ggml-org/whisper.cpp#1461] [Show and tell]
I suddenly wanted to take a closer look at the OpenAI Whisper paper, and one section that caught my attention is the one I highlighted in yellow. Then I checked the whisper.cpp code and found that there are mainly two issues: **_the size of the temperature increment, and the method of calculating the compression ratio._** The temperature in whisper.cpp increases by `0.4` each time instead of the `0.2` mentioned in the paper. Additionally, whisper.cpp uses `entropy` as a substitute for the `gzip compression ratio`, while OpenAI Whisper actually compresses the text and calculates the real `gzip compression ratio`.  @ggerganov 

<img width="380" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/129547291/db1ce789-11aa-4aa1-8613-4b1d105a2445">

> 

### Temprature:
`whisper/transcribe.py`

<img width="477" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/129547291/bdbf0f51-a284-450c-ba18-3bdf6a394591">

> 

https://github.com/ggerganov/whisper.cpp/blob/0de8582f65bde052b08b1844c39fed8808f75619/whisper.cpp#L3833

> 

https://github.com/ggerganov/whisper.cpp/blob/0de8582f65bde052b08b1844c39fed8808f75619/whisper.cpp#L4545-L4554

> 

### Gzip Compression Ratio:
`whisper/utils.py`

<img width="398" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/129547291/0255d502-70fa-466f-96f4-cc820c35f048">

> 

https://github.com/ggerganov/whisper.cpp/blob/0de8582f65bde052b08b1844c39fed8808f75619/whisper.cpp#L4326-L4372

**Comment by ggerganov** [comment]
The reason for the temp increase to be 0.4 is that it is faster processing when the fallback triggers. After we add efficient batched decoding, we will reduce it to 0.2.

We don't use `gzip` because it is not trivial to implement the compression. Entropy estimation works in a similar way and it's much easier to implement

**Reply by bobqianic** [reply]
> We don't use gzip because it is not trivial to implement the compression. Entropy estimation works in a similar way and it's much easier to implement

In some scenarios, this assertion doesn't hold. Consider the following case. While the entropy of this segment is OK, the compression ratio does not meet expectations.

``` 
[best_decoder.sequence.entropy: 3.149363]
[00:33:03.860 --> 00:33:05.860]   I don't know who they are, but I'm sure they're going to tell us something.
[00:33:05.860 --> 00:33:07.860]   I don't know who they are, but I'm sure they're going to tell us something.
[00:33:07.860 --> 00:33:09.860]   I don't know who they are, but I'm sure they're going to tell us something.
[00:33:09.860 --> 00:33:11.860]   I don't know who they are, but I'm sure they're going to tell us something.
[00:33:11.860 --> 00:33:13.860]   I don't know who they are, but I'm sure they're going to tell us something.
[00:33:13.860 --> 00:33:15.860]   I don't know who they are, but I'm sure they're going to tell us something.
[00:33:15.860 --> 00:33:17.860]   I don't know who they are, but I'm sure they're going to tell us something.
[00:33:17.860 --> 00:33:19.860]   I don't know who they are, but I'm sure they're going to tell us something.
[00:33:19.860 --> 00:33:21.860]   I don't know who they are, but I'm sure they're going to tell us something.
```

`gzip` compression ratio is `8.0357` >> `2.4`

```
[_286_][_500_][_380_][_458_][_261_][_1289_][_436_][_366_][_11_][_457_][_286_][_478_][_988_][_436_][_434_][_516_][_281_][_980_][_505_][_746_][_13_]

I don't know who they are, but I'm sure they're going to tell us something.
```

```
[_286_][_500_][_380_][_458_][_567_][_436_][_366_][_11_][_457_][_286_][_478_][_988_][_436_][_434_][_516_][_281_][_980_][_505_][_746_][_13_]

I don't know who they are, but I'm sure they're going to tell us something.
```

**Possible reasons:**
1. **Various combinations of tokens can yield identical text.**
2. **Tokens are created during the training of BPE, and since they are already in a highly compressed form, entropy of tokens doesn't serve as a reliable measure.**
3. **The approach adopted for calculating entropy in whisper.cpp is not sufficient. As highlighted by `C.E. Shannon` in `A Mathematical Theory of Communication`, the formula** $$H(S) = -\sum_{i=1}^{n} p_i \log_2(p_i)$$ **is applicable solely to sequences where successive symbols are independent.**

**Reply by Artoria2e5** [reply]
It could make sense to try a different LZ77 compressor, like the very simple LZ4 here. You only need the match-scanning after all.

**Reply by ggerganov** [reply]
I'm OK to add compression based fallbacks, but the compressor implementation has to be very simple and efficient. I'm not familiar with this topic, so I don't know how much effort it is to implement LZ from scratch.

**Reply by Artoria2e5** [reply]
I will do a little reading of my own on that front. Our requirement can be further broken down into just estimating the compressed-size without coding; one approach would be to read and strip down the source code of an existing tiny compressor like FastLz or LZSS.C, another would be to try and make our own by following parts of https://glinscott.github.io/lz/index.html#toc4.2.2 (no Huffman, no "optimal parsing" even, cuz LZ4 doesn’t do that.)

**Comment by AustinCase** [comment]
Hey @ggerganov and @Artoria2e5 just coming across this thread. Not sure what the current thinking is / timeline of implementation?

**Comment by clort81** [comment]
> I suddenly wanted to take a closer look at the OpenAI Whisper paper, and one section that caught my attention is the one I highlighted in yellow. Then I checked the whisper.cpp code and found that there are mainly two issues: **_the size of the temperature increment, and the method of calculating the compression ratio._** The temperature in whisper.cpp increases by `0.4` each time instead of the `0.2` mentioned in the paper. Additionally, whisper.cpp uses `entropy` as a substitute for the `gzip compression ratio`, while OpenAI Whisper actually compresses the text and calculates the real `gzip compression ratio`. @ggerganov
> <img alt="image" width="380" src="https://private-user-images.githubusercontent.com/129547291/281579887-db1ce789-11aa-4aa1-8613-4b1d105a2445.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk1MzY4OTIsIm5iZiI6MTczOTUzNjU5MiwicGF0aCI6Ii8xMjk1NDcyOTEvMjgxNTc5ODg3LWRiMWNlNzg5LTExYWEtNGFhMS04NjEzLTRiMWQxMDVhMjQ0NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxNFQxMjM2MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NTIyZDYyZWYyYmRlMmQyMTc0MzA0ZTBmZDRkOGRkODc0NTBlY2QzNWFhY2JjNjU4NmUzZTRjMWY0YTZlY2U4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9RzeCVfjroyHEPK68j3DcGoV9O6SpQC-W7i8rqfbwB8">
> 
> > 
> 
> ### Temprature:
> 
> `whisper/transcribe.py`
> <img alt="image" width="477" src="https://private-user-images.githubusercontent.com/129547291/281584064-bdbf0f51-a284-450c-ba18-3bdf6a394591.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk1MzY4OTIsIm5iZiI6MTczOTUzNjU5MiwicGF0aCI6Ii8xMjk1NDcyOTEvMjgxNTg0MDY0LWJkYmYwZjUxLWEyODQtNDUwYy1iYTE4LTNiZGY2YTM5NDU5MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxNFQxMjM2MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDRkZjYzNDNlNzllMmU0NWZlOWUwNmE0Yzc4OTkzNzM0M2IyZGVjYzVjM2FlMGMwYzc0OGNhYWJlNjZhODk0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.zJf2V8aHR6Xco9HAvfBF2pOaKJaDvEVrPDBJfeszxYM">
> 
> > 
> 
> https://github.com/ggerganov/whisper.cpp/blob/0de8582f65bde052b08b1844c39fed8808f75619/whisper.cpp#L3833
> 
> > 
> 
> https://github.com/ggerganov/whisper.cpp/blob/0de8582f65bde052b08b1844c39fed8808f75619/whisper.cpp#L4545-L4554
> 
> > 
> 
> ### Gzip Compression Ratio:
> 
> `whisper/utils.py`
> <img alt="image" width="398" src="https://private-user-images.githubusercontent.com/129547291/281582886-0255d502-70fa-466f-96f4-cc820c35f048.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk1MzY4OTIsIm5iZiI6MTczOTUzNjU5MiwicGF0aCI6Ii8xMjk1NDcyOTEvMjgxNTgyODg2LTAyNTVkNTAyLTcwZmEtNDY2Zi05NmY0LWNjODIwYzM1ZjA0OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxNFQxMjM2MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNTBlOGU1NDkzNTdhZjU4NTIwNWIzZjQ3NjU5ODgzNGNhZmUxMzUxN2ZjNjdkNWRmYjVjOGZmOTNlZDgxNjI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.VvhBfi0dHDj0B1Ph5mWyxx__JC-l4CV5HgjAXbD9bAU">
> 
> > 
> 
> https://github.com/ggerganov/whisper.cpp/blob/0de8582f65bde052b08b1844c39fed8808f75619/whisper.cpp#L4326-L4372

Can it be shown, in whisper.cpp output, that this error also leads to erroneous results, or is it noise below threshold?  
Can that be shown without writing the actual compressor required to compare results to?
